helix-contract/contracts/wallet/MultiSigWallet.sol

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.17;
+
+import {ECDSA} from "@openzeppelin/contracts@4.9.2/utils/cryptography/ECDSA.sol";
+
+contract MultiSigWallet {
+    event ExecutionResult(bytes32 indexed hash, bool result);
+
+    address[] public owners;
+    uint64 public threshold;
+    mapping(address => bool) public isOwner;
+    mapping(bytes32 => bool) public doneOf;
+    mapping(bytes32 => address[]) public transactionSigners;
+
+    receive() external payable {}
+
+    constructor(address[] memory signers, uint64 _threshold) {
+        require(signers.length > 0, "owners required");
+        require(
+            _threshold > 0 && _threshold <= signers.length,
+            "invalid threshold"
+        );
+
+        for (uint256 i = 0; i < signers.length; i++) {
+            address owner = signers[i];
+            require(owner != address(0), "invalid owner");
+            require(!isOwner[owner], "owner not unique");
+
+            isOwner[owner] = true;
+            owners.push(owner);
+        }
+
+        threshold = _threshold;
+    }
+
+    function proposeOrApproveTransaction(
+        address to,
+        uint256 value,
+        uint256 expiration,
+        bytes memory data,
+        bytes calldata signature
+    ) external {
+        require(isOwner[msg.sender], "not an owner");
+        bytes memory txData = abi.encode(
+            block.chainid,
+            address(this),
+            to,
+            value,
+            expiration,
+            data
+        );
+        bytes32 hash = keccak256(txData);
+
+        require(
+            transactionSigners[hash].length < threshold,
+            "threshold reached"
+        );
+        require(
```

Comment:
what do you think of let's directly validate signature here without storing signatures anywhere?
it looks safe to me, only signer with private keys can successfully pass this `require` check
meanwhile, storing less data can save us some gas gee